File: solutions/scala/triangle/1/src/main/scala/Triangle.scala
case class Triangle(a: Double, b: Double, c: Double) {
  private def validTriangle =
    a + b >= c && b + c > a && a + c > b
  
  def equilateral: Boolean = this.validTriangle && a == b && b == c

  def isosceles: Boolean = this.validTriangle && (a == b || a == c || b == c)

  def scalene: Boolean = this.validTriangle && a != b && b != c
}

File: solutions/python/triangle/1/triangle.py
def equilateral(sides):
    [a, b, c] = sides
    return triangle_inequality(a, b, c) and a == b and b == c


def isosceles(sides):
    [a, b, c] = sides
    return triangle_inequality(a, b, c) and (
        a == b or
        a == c  or
        b == c)


def scalene(sides):
    [a, b, c] = sides
    return triangle_inequality(a, b, c) and (
        a != b and b != c and a != c
    )

def triangle_inequality(a, b, c):
    return a > 0 and a + b >= c and b + c >= a and a + c >= b